mesonbuild/modules/selinux.py

```diff
@@ -0,0 +1,344 @@
+                            name = line.split('=', 1)[1].strip().strip('"\'')
+                            break
+            except (IOError, OSError) as e:
+                raise MesonException(f'Failed to read "{config}": {e}')
+
+        mls = T.cast(bool, kwargs.get('mls'))
+        type = T.cast(T.Optional[str], kwargs['type'])
+        if type is None:
+            if name == 'mls':
+                type = 'mls'
+            elif mls:
+                type = 'mcs'
+
+        distro = T.cast(T.Optional[str], kwargs['distro'])
+        direct_initrc = T.cast(T.Optional[bool], kwargs['direct_initrc'])
+
+        headerdir = os.path.join(sharedir, 'devel', 'include')
+        # fallback to system default
+        if not os.path.exists(headerdir):
+            headerdir = '/usr/share/selinux/devel/include'
+        build_conf = os.path.join(headerdir, 'build.conf')
+
+        try:
+            with open(build_conf, 'r', encoding='utf-8') as f:
+                build_vars: T.Dict[str, str] = {}
+                build_vars = {k: v for k, v in {
+                    'TYPE': type,
+                    'NAME': name,
+                    'DISTRO': distro,
+                    'DIRECT_INITRC': 'y' if direct_initrc else 'n'
+                }.items() if v is not None}
+                build_vars = parse_makefile_variables(f.read(), build_vars)
+
+                type = build_vars.get('TYPE', 'standard')
+                name = build_vars['NAME']
+                distro = build_vars['DISTRO']
+                direct_initrc = build_vars.get('DIRECT_INITRC', 'n') == 'y'
+                ubac = build_vars.get('UBAC', 'n') == 'y'
+                mls_sens = build_vars.get('MLS_SENS', '16')
+                mls_cats = build_vars.get('MLS_CATS', '1024')
+                mcs_cats = build_vars.get('MCS_CATS', '1024')
+        except (IOError, OSError) as e:
+            raise MesonException(f'Failed to read "{build_conf}": {e}')
+
+        m4 = tools['m4']
+        checkmodule = tools['checkmodule']
+        semodule_package = tools['semodule_package']
+
+        m4_args: T.List[str] = []
+        if type == 'mls':
+            m4_args.extend(['-D', 'enable_mls'])
+        elif type == 'mcs':
+            m4_args.extend(['-D', 'enable_mcs'])
+        if distro:
+            m4_args.extend(['-D', f'distro_{distro}'])
+        if direct_initrc:
+            m4_args.extend(['-D', 'direct_sysadm_daemon'])
+        if ubac:
+            m4_args.extend(['-D', 'enable_ubac'])
+
+        m4_args.extend(['-D', 'hide_broken_symptoms'])
+        m4_args.extend(['-D', f'mls_num_sens={mls_sens}'])
+        m4_args.extend(['-D', f'mls_num_cats={mls_cats}'])
+        m4_args.extend(['-D', f'mcs_num_cats={mcs_cats}'])
+
+        # gather the support files
+        m4supportdir = os.path.join(headerdir, 'support')
+        m4support = glob.glob(os.path.join(m4supportdir, '*.spt'))
+
+        # gather all the {headerdir}/*/*.if files, except if the subdir is 'support'
```

Comment:
Meson does not support globs by design, and this code demonstrates exactly why: it gets parsed at configure time and then changes to the glob results fail to trigger a reconfigure. Do not do this: implement the globbing as part of the custom_target command instead, and emit a depfile as a side effect of compilation. I presume that the addition of a new file isn't *supposed* to affect anything unless an existing file is updated to declare a dependency on the new file or the in-tree policy is updated to declare a dependency on the new file (this is, in broad approach, how C headers work too, of course).

Reply:
done, I moved most of the logic to the script, where a depfile is generated.